nexus/src/app/sled.rs

```diff
@@ -231,4 +233,166 @@ impl super::Nexus {
         self.db_datastore.service_upsert(opctx, service).await?;
         Ok(())
     }
+
+    // OPTE V2P mappings
+
+    /// Ensure that the necessary v2p mappings for an instance are created
+    pub async fn create_instance_v2p_mappings(
+        &self,
+        opctx: &OpContext,
+        instance_id: Uuid,
+    ) -> Result<(), Error> {
+        // For every sled that isn't the sled this instance was allocated to, create
+        // a virtual to physical mapping for each of this instance's NICs.
+        //
+        // TODO this is a first-pass naive approach. A more optimized approach would
+        // be to see what other instances share the VPC this instance is in (or more
+        // generally, what instances should have connectivity to this one), see what
+        // sleds those are allocated to, and only create a v2p mapping for those
+        // sleds. However, this naive approach has the advantage of not requiring
+        // additional work: v2p mappings have to be bidirectional in order for both
+        // instances's packets to make a round trip. After this function executes, a
+        // v2p mapping to this instance will exist on every sled, and when another
+        // instance is allocated to a random sled, that sled will already contain a
+        // mapping to this instance.
+        //
+        // TODO-correctness OPTE currently will block instances in different VPCs
+        // from connecting to each other. If it ever stops doing this, this naive
+        // approach will create v2p mappings that shouldn't exist.
+
+        let (.., authz_instance, db_instance) =
+            LookupPath::new(&opctx, &self.db_datastore)
+                .instance_id(instance_id)
+                .fetch_for(authz::Action::Read)
+                .await?;
+
+        let instance_nics = self
+            .db_datastore
+            .derive_guest_network_interface_info(&opctx, &authz_instance)
+            .await?;
+
+        // Lookup the physical host IP of the sled hosting this instance
+        let instance_sled_id = db_instance.runtime().sled_id;
+        let physical_host_ip: std::net::IpAddr =
+            (*self.sled_lookup(&self.opctx_alloc, &instance_sled_id).await?.ip)
+                .into();
+
+        let mut last_sled_id: Option<Uuid> = None;
+        loop {
```

Comment:
Maybe a nit, bu it seems like this loop could be parallelized and then collected, via something like `FuturesUnordered`.

Reply:
I wasn't smart enough to do this with `FuturesUnordered`, so I did it with `tokio::spawn` instead: 117f8b30